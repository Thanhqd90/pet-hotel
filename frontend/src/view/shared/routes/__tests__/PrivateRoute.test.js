import { mount } from 'enzyme';
import React, { Component } from 'react';
import {
  BrowserRouter,
  Link,
  Route,
  Switch,
} from 'react-router-dom';
import PrivateRoute from 'view/shared/routes/PrivateRoute';
import Roles from 'security/roles';
const roles = Roles.values;

function mountPage(currentUser, permissionRequired) {
  const PrivatePage = class PrivatePage extends Component {
    render() {
      return (
        <div data-test="PrivatePage">Private Page</div>
      );
    }
  };

  const SignInPage = class SignInPage extends Component {
    render() {
      const { from } = this.props.location.state || {
        from: { pathname: '/' },
      };

      return (
        <div data-test="SignInPage">
          Singin Page
          <div data-test-from={from.pathname} />
        </div>
      );
    }
  };

  return mount(
    <BrowserRouter>
      <div>
        <Link to="/private">Go to private</Link>

        <Switch>
          <Route
            exact
            path="/auth/signin"
            component={SignInPage}
          />

          <Route
            exact
            path="/auth/email-unverified"
            render={() => (
              <div data-test="EmailUnverifiedPage">
                Email Unverified
              </div>
            )}
          />

          <Route
            exact
            path="/auth/empty-permissions"
            render={() => (
              <div data-test="EmptyPermissionsPage">
                Empty Permissions
              </div>
            )}
          />

          <Route
            exact
            path="/403"
            render={() => <div data-test="403">403</div>}
          />

          <PrivateRoute
            path="/private"
            currentUser={currentUser}
            permissionRequired={permissionRequired}
            component={PrivatePage}
          />
        </Switch>
      </div>
    </BrowserRouter>,
  );
}

describe('view/shared/routes/PrivateRoute', () => {
  describe('Not Authenticated', () => {
    let component;
    beforeAll(() => {
      component = mountPage(null, []);

      component
        .find('Link[to="/private"]')
        .simulate('click', { button: 0 });
    });

    test('redirect to login', () => {
      expect(component.html()).toMatchSnapshot();

      expect(
        component
          .find(`[data-test="PrivatePage"]`)
          .exists(),
      ).toBeFalsy();

      expect(
        component.find(`[data-test="SignInPage"]`).exists(),
      ).toBeTruthy();
    });

    test('send from to signin', () => {
      expect(
        component
          .find(`[data-test-from="/private"]`)
          .exists(),
      ).toBeTruthy();
    });
  });

  describe('Email Unverified', () => {
    let component;

    beforeAll(() => {
      component = mountPage(
        {
          id: 1,
          roles: [roles.owner],
          emailVerified: false,
        },
        { allowedRoles: [roles.owner] },
      );

      component
        .find('Link[to="/private"]')
        .simulate('click', { button: 0 });
    });

    test('redirects to email unverified', () => {
      expect(component.html()).toMatchSnapshot();

      expect(
        component
          .find(`[data-test="EmailUnverifiedPage"]`)
          .exists(),
      ).toBeTruthy();
    });
  });

  describe('Empty Permissions', () => {
    let component;

    beforeAll(() => {
      component = mountPage(
        {
          id: 1,
          roles: [],
          emailVerified: true,
        },
        { allowedRoles: [roles.owner] },
      );

      component
        .find('Link[to="/private"]')
        .simulate('click', { button: 0 });
    });

    test('redirects to empty permissions', () => {
      expect(component.html()).toMatchSnapshot();

      expect(
        component
          .find(`[data-test="EmptyPermissionsPage"]`)
          .exists(),
      ).toBeTruthy();
    });
  });

  describe('Not enough permissions', () => {
    let component;

    beforeAll(() => {
      component = mountPage(
        {
          id: 1,
          roles: [roles.viewer],
          emailVerified: true,
        },
        { allowedRoles: [roles.owner] },
      );

      component
        .find('Link[to="/private"]')
        .simulate('click', { button: 0 });
    });

    test('redirects to 403', () => {
      expect(component.html()).toMatchSnapshot();

      expect(
        component.find(`[data-test="403"]`).exists(),
      ).toBeTruthy();
    });
  });

  describe('Authenticated with emailVerified and roles', () => {
    let component;

    beforeAll(() => {
      component = mountPage(
        {
          id: 1,
          roles: [roles.owner],
          emailVerified: true,
        },
        { allowedRoles: [roles.owner] },
      );

      component
        .find('Link[to="/private"]')
        .simulate('click', { button: 0 });
    });

    test('opens the page', () => {
      expect(component.html()).toMatchSnapshot();

      expect(
        component
          .find(`[data-test="PrivatePage"]`)
          .exists(),
      ).toBeTruthy();
    });
  });
});
