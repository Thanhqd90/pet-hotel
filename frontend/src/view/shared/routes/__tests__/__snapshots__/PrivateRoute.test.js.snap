// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view/shared/routes/PrivateRoute Authenticated with emailVerified and roles opens the page 1`] = `"<div><a href=\\"/private\\">Go to private</a><div data-test=\\"PrivatePage\\">Private Page</div></div>"`;

exports[`view/shared/routes/PrivateRoute Email Unverified redirects to email unverified 1`] = `"<div><a href=\\"/private\\">Go to private</a><div data-test=\\"EmailUnverifiedPage\\">Email Unverified</div></div>"`;

exports[`view/shared/routes/PrivateRoute Empty Permissions redirects to empty permissions 1`] = `"<div><a href=\\"/private\\">Go to private</a><div data-test=\\"EmptyPermissionsPage\\">Empty Permissions</div></div>"`;

exports[`view/shared/routes/PrivateRoute Not Authenticated redirect to login 1`] = `"<div><a href=\\"/private\\">Go to private</a><div data-test=\\"SignInPage\\">Singin Page<div data-test-from=\\"/private\\"></div></div></div>"`;

exports[`view/shared/routes/PrivateRoute Not enough permissions redirects to 403 1`] = `"<div><a href=\\"/private\\">Go to private</a><div data-test=\\"403\\">403</div></div>"`;
